tapir/shifts/views.py

```diff
@@ -700,6 +703,35 @@ def get_target_user_data(self) -> ShiftUserData:
 
     def form_valid(self, form):
         form.instance.shift_user_data = self.get_target_user_data()
+        exemption: ShiftExemption = form.instance
+        user = self.get_target_user_data().user
+        for attendance in ShiftExemption.get_attendances_covered_by_exemption(
+            user=user,
+            start_date=exemption.start_date,
+            end_date=exemption.end_date,
+        ):
+            attendance.state = ShiftAttendance.State.CANCELLED
+            attendance.excused_reason = (
+                _("Is covered by shift exemption: ") + exemption.description
+            )
+            attendance.save()
+
+        if ShiftExemption.must_unregister_from_abcd_shift(
+            start_date=exemption.start_date, end_date=exemption.end_date
+        ):
+            for attendance_template in ShiftAttendanceTemplate.objects.filter(
+                user=user
+            ):
+                attendance_template.cancel_attendances(
```

Comment:
There are two different codepaths to do essentially the same thing here: Cancel all attendances in a date range.

I feel `cancel_attendances` should take a date range tuple (or start_date, end_date, whatever you prefer). Both paths should use this method.

Maybe even add a `ShiftAttendanceTemplate.delete_and_cancel_future_attendances` since we're doing it in two different places?

Things like this make it obvious that Django is really missing a controller layer (or whatever you want to call it) between the model and the view...

Reply:
On the first path (L708 to 717) there's no `ShiftAttendanceTemplate` to call cancel_attendances from, so instead I changed the function to `get_attendances_cancelled_by_exemption()` so that we get all the attendances from both pathes in one set.